File: ios/Orcha/Domain/DeviceAuth.swift
import Foundation

/// The QR → GitHub OAuth device-token contract (orcha-cloud): the pure URL
/// logic on both edges of the browser round-trip, kept free of session/UI
/// concerns so it is directly testable.
///
/// Outbound: `https://<host>/oauth2/start?rd=%2Fauth%2Fdevice` — oauth2-proxy
/// runs GitHub OAuth, then the authenticated portal page `/auth/device` mints
/// a per-device token. Inbound: that page redirects to
/// `orcha://auth/callback?host=<host>&token=<token>`, which the
/// ASWebAuthenticationSession intercepts and hands back as a URL.
enum DeviceAuth {
    /// The registered custom scheme the auth session watches for.
    static let callbackScheme = "orcha"

    /// The minted credential handed back by the portal's device page.
    struct Callback: Equatable {
        let host: String
        let token: String
    }

    /// The browser entry point for a protected deployment: its oauth2-proxy
    /// start URL, returning to the portal's device-token page. Built as a
    /// string so `rd` stays exactly `%2Fauth%2Fdevice` — the encoded form the
    /// proxy expects, which URLComponents would leave as literal slashes.
    static func startURL(forBase base: String) -> URL? {
        URL(string: base + "/oauth2/start?rd=%2Fauth%2Fdevice")
    }

    /// True for any `orcha://auth/...` URL — the sub-namespace owned by this
    /// flow. `onOpenURL` must swallow these rather than route them as deep
    /// links: the live session consumes the callback itself, so one arriving
    /// at the app is a stray (stale browser tab after the session closed).
    static func isAuthCallback(_ url: URL) -> Bool {
        url.scheme?.lowercased() == callbackScheme && url.host?.lowercased() == "auth"
    }

    /// Parse `orcha://auth/callback?host=<host>&token=<token>`; nil for
    /// anything else — wrong scheme/host/path, missing or empty parameters.
    static func parseCallback(_ url: URL) -> Callback? {
        guard isAuthCallback(url), url.path == "/callback",
              let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems
        else { return nil }
        func value(_ name: String) -> String? {
            let raw = items.first { $0.name == name }?.value?
                .trimmingCharacters(in: .whitespacesAndNewlines)
            return (raw?.isEmpty ?? true) ? nil : raw
        }
        guard let host = value("host"), let token = value("token") else { return nil }
        return Callback(host: host, token: token)
    }

    /// The minted token is only ever attached to the deployment the flow
    /// started against: the callback's `host` must name the same host as the
    /// pending base URL. Lenient about the shapes a server might send —
    /// bare host, host:port, or a full URL — strict about the host itself.
    static func callback(_ callback: Callback, matchesBase base: String) -> Bool {
        guard let baseHost = URL(string: base)?.host?.lowercased() else { return false }
        let raw = callback.host.lowercased()
        let candidate = raw.contains("://")
            ? URL(string: raw)?.host
            : raw.split(separator: ":").first.map(String.init)
        return candidate == baseHost
    }
}

File: android/app/src/main/java/io/openorcha/mobile/domain/DeviceAuth.kt
package io.openorcha.mobile.domain

import java.net.URI
import java.net.URLDecoder

/**
 * The QR -> GitHub OAuth device-token contract (orcha-cloud): the pure URL logic on
 * both edges of the browser round-trip, kept free of session/UI concerns so it is
 * directly testable. Ports `ios/Orcha/Domain/DeviceAuth.swift` 1:1. Deliberately
 * built on `java.net.URI` rather than `android.net.Uri` -- like `OrchaServerAddress`
 * -- so this stays a plain-JUnit unit under `src/test`, no Robolectric/instrumentation
 * needed.
 *
 * Outbound: `https://<host>/auth/device` -- the perimeter walks the user through
 * GitHub OAuth, then the authenticated portal page mints a per-device token.
 * Inbound: that page redirects to `orcha://auth/callback?host=<host>&token=<token>`,
 * which the Custom Tab intent-filter on `MainActivity` intercepts.
 */
object DeviceAuth {
    /** The registered custom scheme the callback intent-filter watches for. */
    const val CALLBACK_SCHEME = "orcha"

    /** The minted credential handed back by the portal's device page. */
    data class Callback(val host: String, val token: String)

    /**
     * The browser entry point for a protected deployment: its device-token sign-in
     * page. `base` is the pairing base URL, e.g. `https://orcha.quantallabs.ai`.
     */
    fun startUrl(base: String): String? {
        val trimmed = base.trim().trimEnd('/')
        if (trimmed.isBlank()) return null
        return "$trimmed/auth/device"
    }

    /**
     * True for any `orcha://auth/...` URI -- the sub-namespace owned by this flow.
     * Any stray delivery of one of these (e.g. a relaunch from the recents tray)
     * must be swallowed rather than routed as a normal deep link.
     */
    fun isAuthCallback(uri: URI): Boolean =
        uri.scheme?.lowercase() == CALLBACK_SCHEME && uri.host?.lowercase() == "auth"

    /**
     * Parse `orcha://auth/callback?host=<host>&token=<token>`; null for anything
     * else -- wrong scheme/host/path, missing or empty parameters.
     */
    fun parseCallback(uri: URI): Callback? {
        if (!isAuthCallback(uri)) return null
        if (uri.path != "/callback") return null
        val params = parseQuery(uri.rawQuery)
        val host = params["host"]?.trim()?.takeIf { it.isNotEmpty() } ?: return null
        val token = params["token"]?.trim()?.takeIf { it.isNotEmpty() } ?: return null
        return Callback(host, token)
    }

    /** Parse a raw `orcha://...` string; null on any malformed URI. */
    fun parseCallback(raw: String): Callback? =
        runCatching { URI(raw) }.getOrNull()?.let { parseCallback(it) }

    /**
     * The minted token is only ever attached to the deployment the flow started
     * against: the callback's `host` must name the same host as the pending base
     * URL. Lenient about the shapes a server might send -- bare host, host:port,
     * or a full URL -- strict about the host itself.
     */
    fun callbackMatchesBase(callback: Callback, base: String): Boolean {
        val baseHost = runCatching { URI(base).host }.getOrNull()?.lowercase() ?: return false
        val raw = callback.host.lowercase()
        val candidate = if (raw.contains("://")) {
            runCatching { URI(raw).host }.getOrNull()
        } else {
            raw.split(":").firstOrNull()
        }
        return candidate == baseHost
    }

    /** Minimal `a=1&b=2` decoder -- avoids pulling in a URI-component library. */
    private fun parseQuery(rawQuery: String?): Map<String, String> {
        if (rawQuery.isNullOrEmpty()) return emptyMap()
        return rawQuery.split("&").mapNotNull { pair ->
            if (pair.isEmpty()) return@mapNotNull null
            val idx = pair.indexOf('=')
            val key = if (idx >= 0) pair.substring(0, idx) else pair
            val value = if (idx >= 0) pair.substring(idx + 1) else ""
            val decodedKey = runCatching { URLDecoder.decode(key, "UTF-8") }.getOrDefault(key)
            val decodedValue = runCatching { URLDecoder.decode(value, "UTF-8") }.getOrDefault(value)
            decodedKey to decodedValue
        }.toMap()
    }
}
